File: sdk/php/tests/e2e/rest/Account/AccountFeeTest.php
<?php

namespace Tests\e2e\rest\Account;

use JMS\Serializer\Serializer;
use JMS\Serializer\SerializerBuilder;
use KuCoin\UniversalSDK\Api\DefaultClient;
use KuCoin\UniversalSDK\Common\Logger;
use KuCoin\UniversalSDK\Generate\Account\Fee\FeeApi;
use KuCoin\UniversalSDK\Generate\Account\Fee\GetBasicFeeReq;
use KuCoin\UniversalSDK\Generate\Account\Fee\GetFuturesActualFeeReq;
use KuCoin\UniversalSDK\Generate\Account\Fee\GetSpotActualFeeReq;
use KuCoin\UniversalSDK\Internal\Utils\JsonSerializedHandler;
use KuCoin\UniversalSDK\Model\ClientOptionBuilder;
use KuCoin\UniversalSDK\Model\Constants;
use KuCoin\UniversalSDK\Model\TransportOptionBuilder;
use PHPUnit\Framework\TestCase;

class AccountFeeTest extends TestCase
{
    /**
     * @var FeeApi $api
     */
    private $api;

    /**
     * @var Serializer $serializer
     */
    private $serializer;

    protected function setUp(): void
    {

        $this->serializer = SerializerBuilder::create()
            ->addDefaultHandlers()
            ->configureHandlers(function ($handlerRegistry) {
                $handlerRegistry->registerSubscribingHandler(
                    new JsonSerializedHandler()
                );
            })
            ->build();

        // Retrieve API secret information from environment variables
        $key = getenv('API_KEY') ?: '';
        $secret = getenv('API_SECRET') ?: '';
        $passphrase = getenv('API_PASSPHRASE') ?: '';

        // Optional: Retrieve broker secret information from environment variables; applicable for broker API only
        $brokerName = getenv('BROKER_NAME');
        $brokerKey = getenv('BROKER_KEY');
        $brokerPartner = getenv('BROKER_PARTNER');

        // Set specific options, others will fall back to default values
        $httpTransportOption = (new TransportOptionBuilder())
            ->setKeepAlive(true)
            ->setMaxConnections(10)
            ->build();

        // Create a client using the specified options
        $clientOption = (new ClientOptionBuilder())
            ->setKey($key)
            ->setSecret($secret)
            ->setPassphrase($passphrase)
            ->setBrokerName($brokerName)
            ->setBrokerKey($brokerKey)
            ->setBrokerPartner($brokerPartner)
            ->setSpotEndpoint(Constants::GLOBAL_API_ENDPOINT)
            ->setFuturesEndpoint(Constants::GLOBAL_FUTURES_API_ENDPOINT)
            ->setBrokerEndpoint(Constants::GLOBAL_BROKER_API_ENDPOINT)
            ->setTransportOption($httpTransportOption)
            ->build();

        $client = new DefaultClient($clientOption);
        $kucoinRestService = $client->restService();
        $this->api = $kucoinRestService->getAccountService()->getFeeApi();
    }


    /**
     * getBasicFee
     * Get Basic Fee - Spot/Margin
     * /api/v1/base-fee
     */
    public function testGetBasicFee()
    {
        $builder = GetBasicFeeReq::builder();
        $builder->setCurrencyType(0);
        $req = $builder->build();
        $resp = $this->api->getBasicFee($req);
        self::assertNotNull($resp->takerFeeRate);
        self::assertNotNull($resp->makerFeeRate);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * getSpotActualFee
     * Get Actual Fee - Spot/Margin
     * /api/v1/trade-fees
     */
    public function testGetSpotActualFee()
    {
        $builder = GetSpotActualFeeReq::builder();
        $builder->setSymbols('BTC-USDT,ETH-USDT');
        $req = $builder->build();
        $resp = $this->api->getSpotActualFee($req);
        foreach ($resp->data as $item) {
            self::assertNotNull($item->symbol);
            self::assertNotNull($item->takerFeeRate);
            self::assertNotNull($item->makerFeeRate);
        }

        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * getFuturesActualFee
     * Get Actual Fee - Futures
     * /api/v1/trade-fees
     */
    public function testGetFuturesActualFee()
    {
        $builder = GetFuturesActualFeeReq::builder();
        $builder->setSymbol("XBTUSDM");
        $req = $builder->build();
        $resp = $this->api->getFuturesActualFee($req);
        self::assertNotNull($resp->symbol);
        self::assertNotNull($resp->takerFeeRate);
        self::assertNotNull($resp->makerFeeRate);
        Logger::info($resp->jsonSerialize($this->serializer));
    }


}

File: sdk/java/src/test/java/com/kucoin/universal/sdk/test/e2e/rest/account/AccountFeeTest.java
package com.kucoin.universal.sdk.test.e2e.rest.account;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.kucoin.universal.sdk.api.DefaultKucoinClient;
import com.kucoin.universal.sdk.api.KucoinClient;
import com.kucoin.universal.sdk.generate.account.fee.*;
import com.kucoin.universal.sdk.model.ClientOption;
import com.kucoin.universal.sdk.model.Constants;
import com.kucoin.universal.sdk.model.TransportOption;
import java.io.IOException;
import java.util.Collections;
import lombok.extern.slf4j.Slf4j;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

@Slf4j
public class AccountFeeTest {

  private static FeeApi api;

  public static ObjectMapper mapper = new ObjectMapper();

  @BeforeAll
  public static void setUp() {

    String key = System.getenv("API_KEY");
    String secret = System.getenv("API_SECRET");
    String passphrase = System.getenv("API_PASSPHRASE");

    TransportOption httpTransport =
        TransportOption.builder()
            .interceptors(
                Collections.singleton(
                    new Interceptor() {
                      @NotNull @Override
                      public Response intercept(@NotNull Chain chain) throws IOException {
                        Request request = chain.request();

                        System.out.println("========== Request ==========");
                        System.out.println(request.method() + " " + request.url());

                        Response response = chain.proceed(request);

                        System.out.println("========== Response ==========");
                        System.out.println("Status Code: " + response.code());
                        System.out.println("Message: " + response.message());
                        return response;
                      }
                    }))
            .build();

    ClientOption clientOpt =
        ClientOption.builder()
            .key(key)
            .secret(secret)
            .passphrase(passphrase)
            .spotEndpoint(Constants.GLOBAL_API_ENDPOINT)
            .futuresEndpoint(Constants.GLOBAL_FUTURES_API_ENDPOINT)
            .brokerEndpoint(Constants.GLOBAL_BROKER_API_ENDPOINT)
            .transportOption(httpTransport)
            .build();

    KucoinClient kucoinClient = new DefaultKucoinClient(clientOpt);
    api = kucoinClient.getRestService().getAccountService().getFeeApi();
  }

  /** getBasicFee Get Basic Fee - Spot/Margin /api/v1/base-fee */
  @Test
  public void testGetBasicFee() throws Exception {
    GetBasicFeeReq.GetBasicFeeReqBuilder builder = GetBasicFeeReq.builder();
    builder.currencyType(GetBasicFeeReq.CurrencyTypeEnum._0);
    GetBasicFeeReq req = builder.build();
    GetBasicFeeResp resp = api.getBasicFee(req);
    Assertions.assertNotNull(resp.getTakerFeeRate());
    Assertions.assertNotNull(resp.getMakerFeeRate());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** getSpotActualFee Get Actual Fee - Spot/Margin /api/v1/trade-fees */
  @Test
  public void testGetSpotActualFee() throws Exception {
    GetSpotActualFeeReq.GetSpotActualFeeReqBuilder builder = GetSpotActualFeeReq.builder();
    builder.symbols("BTC-USDT,ETH-USDT");
    GetSpotActualFeeReq req = builder.build();
    GetSpotActualFeeResp resp = api.getSpotActualFee(req);
    resp.getData()
        .forEach(
            item -> {
              Assertions.assertNotNull(item.getSymbol());
              Assertions.assertNotNull(item.getTakerFeeRate());
              Assertions.assertNotNull(item.getMakerFeeRate());
            });

    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** getFuturesActualFee Get Actual Fee - Futures /api/v1/trade-fees */
  @Test
  public void testGetFuturesActualFee() throws Exception {
    GetFuturesActualFeeReq.GetFuturesActualFeeReqBuilder builder = GetFuturesActualFeeReq.builder();
    builder.symbol("XBTUSDM");
    GetFuturesActualFeeReq req = builder.build();
    GetFuturesActualFeeResp resp = api.getFuturesActualFee(req);
    Assertions.assertNotNull(resp.getSymbol());
    Assertions.assertNotNull(resp.getTakerFeeRate());
    Assertions.assertNotNull(resp.getMakerFeeRate());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }
}
